core/src/main/scala/chisel3/SimpleChiselTransformer.scala

```diff
@@ -6,91 +6,178 @@ import chisel3.internal.firrtl._
+              arg_buf += arg
             }
+          }
+          ctx.update(
+            i,
+            DefPrim(dPrim.sourceInfo, dPrim.id, dPrim.op, arg_buf.toSeq: _*)
+          )
         }
-        return ctx
-    }
-
-    //replace lId with rId in ctx and return the new ctx
-    def replaceWhen(lId: Arg, rId: Arg, ctx: ArrayBuffer[Command])(pred: () => Boolean): Option[ArrayBuffer[Command]] ={
-        if(pred()){
-            val new_ctx = new ArrayBuffer[Command]
-            return Some(new_ctx)
+        case regInit: DefRegInit => {
+          if (regInit.init.uniqueId.equals(lId.uniqueId)) regInit.init = rId.ref
         }
-        else{
-            return Some(ctx)
+        case whenBegin: WhenBegin => {
+          if (whenBegin.pred.uniqueId.equals(lId.uniqueId))
+            whenBegin.pred = rId.ref
         }
+        case connect: Connect => {
+          if (connect.exp.uniqueId.equals(lId.uniqueId)) connect.exp = rId.ref
+        }
+        case connect: ConnectInit => {
+          if (connect.exp.uniqueId.equals(lId.uniqueId)) connect.exp = rId.ref
+        }
+        case _ => ()
+      }
+    }
+    return ctx
+  }
+
+  //replace lId with rId in ctx and return the new ctx
+  def replaceWhen(lId: Arg, rId: Arg, ctx: ArrayBuffer[Command])(
+      pred: () => Boolean
+  ): Option[ArrayBuffer[Command]] = {
+    if (pred()) {
+      val new_ctx = new ArrayBuffer[Command]
+      return Some(new_ctx)
+    } else {
+      return Some(ctx)
+    }
+  }
+
+  private def dependsOnArg(cmd: Command, eql: Arg => Boolean): Boolean =
+    cmd match {
+      case Connect(_, _, exp)     => eql(exp)
+      case WhenBegin(_, pred)     => eql(pred)
+      case ConnectInit(_, _, exp) => eql(exp)
+      case Stop(_, clock, _)      => eql(clock)
+      case Printf(_, clock, _)    => eql(clock)
+      case _                      => false
     }
 
-    def removeIfSafe(id: Arg, ctx: ArrayBuffer[Command]): Option[ArrayBuffer[Command]] ={
-        //TODO
-        return None   
+  private def getDefinitionUid(command: Command): Option[BigInt] =
+    command match {
+      // Right(args.find(a => eql(a)).map(a => a.uniqueId)) what's the id here?
+      case DefPrim(_, _, _, args @ _*)  => Some(args(0).uniqueId)
+      case DefInvalid(_, arg)           => Some(arg.uniqueId)
+      case DefWire(_, data)             => Some(data.ref.uniqueId)
+      case DefReg(_, data, _)           => Some(data.ref.uniqueId)
+      case DefRegInit(_, data, _, _, _) => Some(data.ref.uniqueId)
+      case _                            => None
     }
 
-    // find the most front place to insert the cmd
-    def Insert(cmd: Command, ctx: ArrayBuffer[Command]): Option[ArrayBuffer[Command]] ={
-        //TODO
-        return None   
+  def removeIfSafe(
+      id: Arg,
+      ctx: ArrayBuffer[Command]
+  ): Option[ArrayBuffer[Command]] = {
+
+    def argEquality(a: Arg) = a.uniqueId.equals(id.uniqueId)
+
+    def matchesUID(c: Command) = getDefinitionUid(c) match {
```

Comment:
This doesn't feel idiomatic to me. basically want the truth value of the Option when the function is applied to it. Might be better done with a map like this: 

`      getDefinitionUid(c).map(a => a.equals(id.uniqueId)).getOrElse(false)
`